Impute Constant

gracetian6 opened the pull request:
`make_impute_constant` is a function that "imputes" or inserts a `constant` input variable whenever the vector entry is null. 

Changes as of July 26:

- pseudocode with preconditions and post conditions
- proofs with appropriate output domain, domain metric, stability guarantee

**PRIORITIZE: Appropriate Output Domain Proof.** Not really sure if checking type signature is enough? Unsure whether the error check in the code needs to be included in pseudo code (and is needed to justify output domain pf).

Please look at [impute_constant.pdf](https://github.com/opendp/whitepapers/blob/impute_constant/impute_constant/impute_constant.pdf) as you review the .tex!

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,158 @@
+\end{theorem}
+\begin{proof}
+\begin{enumerate}
+\item \textbf{(Appropriate output domain).} 
+In the case of \texttt{make\_impute\_constant}, this corresponds to showing that for every vector $v$ of elements of type \texttt{DA}, \texttt{function}$(v)$ is a vector of elements of type \texttt{DA::NonNull}.
+
+The $\function(v)$ has type \texttt{Vec(DA)} follows from the assumption that element $v$ is in \texttt{input\_domain} and from the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}), which takes in an element of type \texttt{Vec(DA)} and returns an element of type \texttt{Vec(DA::NonNull)}. If the Rust code compiles correctly, then the type ...
+
+Lastly, we ensure that return type must be \texttt{NonNull} \grace{Attribute? Property? Carrier} \texttt{DA::NonNull} because we check to make sure the \texttt{constant} being imputed in the function must be null. \grace{Not sure if this needs to be explicit in the pseudo code with the error check, or if I can just cite some trait property.} This check already exists because in the input of the \ ...
```
Want feedback on appropriate output domain - not sure how to make it more rigorous, and if I need to consider more than just the type signature.

Shoeboxam replied:
[Based on this message](https://github.com/opendp/whitepapers/pull/8#issuecomment-889511572) I think there should first be a definition of AllDomain<T> along the lines of "the set of all non-null values that T may take on." You want to show that the output domain on this constructor is `VectorDomain<AllDomain<T>>`, or "any `Vec<T>` without nulls". You can first lean on the type system to get you most of the way there- the type system constrains the output set to "any `Vec<T>`". You can then show that the set of potential outputs is equivalent to the set described by `VectorDomain<AllDomain<T>>` by showing that vectors with nulls are not members of the output set.

Shoeboxam replied:
Unmarking this as resolved, because you resolved it before I finished writing my response!

Shoeboxam replied:
The Ord changes make this comment is partially outdated. AllDomain will not be adjusted.

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,158 @@
+    \item Either a valid \texttt{Transformation} is returned or an error is returned.
+\end{itemize}
+
+\grace{Not sure if I need to include the error check for whether the constant is nonnull in line 69 code. Is it already checked? }
```
In picture in pdf, in[ line 11 in the Rust code ](https://github.com/opendp/opendp/blob/21-impute/rust/opendp/src/trans/impute.rs#L62-L75) has an error check. Not sure whether / how to include it in pseudocode.

gracetian6 commented:
Changes as of Thu Jul 29:
- incorporated check for nullity in the pseudo code (per Mike's comments)
- added lemma that shows that DA::NonNull can contain null values
- updated proof for appropriate output domain to take into account the check for nullity (of the lemma above)

PRIORITIZE (see my comments on the code below):
1. In the **pseudocode**: Is the type signature correct (input and output type)? Are the input and output domain correct?
2. In the **appropriate output domain proof**: What is the reason the type signature is not sufficient? Is it because the output type is more general than the output domain? Or is it theoretically because constant can be null value even if it has the type DA::NonNull? 
3. **First lemma in proof**: so, is it correct that a variable with type DA::NonNull can contain null values?

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,181 @@
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*1
+
+    def function(data: Vec[DA::Carrier]) -> Vec[DA::NonNull]: |\label{line:fn}|
```
1. is this type signature correct for the function in line 123

Shoeboxam replied:
The function's type signature looks mechanically correct. We already have DA, the Atomic Domain, the domain inside VectorDomain.
- the input type is a vector of DA's associated "Carrier" type
- the output type is a vector of DA's associated "NonNull" type; aka the type after imputation

I'm curious what opinions people have on syntax here.

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,181 @@
+\begin{lstlisting}[language=Python,  escapechar=|]
+def make_impute_constant(constant : DA::NonNull):
+    # instead of VectorDomain(DA), we add ::new() to get a new instance of DA. This is bcause DA has the ImputableDomain trait. Discuss among interns.
+    input_domain = VectorDomain(DA::new())
+    output_domain = VectorDomain(AllDomain(DA::NonNull))
```
1. Is this input and output domain correct?

Shoeboxam replied:
Again, this looks mechanically correct and consistent with the other proofs in that you're moving type arguments into the constructor arguments (line 114). You're making an instance of DA on line 113, but in line 114 passing the type itself in. This is right, but maybe unavoidably a bit ugly.

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,181 @@
+
+The $\function(v)$ has type \texttt{Vec(DA)} follows from the assumption that element $v$ is in \texttt{input\_domain} and from the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}), which takes in an element of type \texttt{Vec(DA)} and returns an element of type \texttt{Vec(DA::NonNull)}. If the Rust code compiles correctly, then the type ...
+
+Lemma \ref{lemma-NonNull} tells us the type signature is not sufficient check because (1) the output type of \texttt{VectorDomain(DA::NonNull)} is more general than the output domain of \texttt{VectorDomain(DA::NonNull)}. \grace{?? The output type matches the output domain I wrote, so something's wrong.} (2) \texttt{constant} can be \texttt{null} even if it is of type \texttt{DA::NonNull}. For (2 ...
```
2. for appropriate output domain: 

What is the reason the type signature is not sufficient? Is it because the output type is more general than the output domain? Or is it theoretically because constant can be null value even if it has the type DA::NonNull?

Shoeboxam replied:
Hmm, `Vec(DA)` or "a vector of atomic domains" doesn't really make sense in this context. I think you meant `Vec(DA::Carrier)` in these blocks instead of `Vec(DA)`. 

And yes, in this case the output type can represent a value (float nan) that is not a member of the output domain. The purpose of the impute function is to ensure that these outputs are never null at runtime, so it replaces null with a constant. Since `constant` has type DA::NonNull, and DA::NonNull may itself be f64, we must check that `constant` is non-null. This check should make it impossible for analysts to "impute" by nonsensically replacing nulls with nulls, thus violating the output domain.

gracetian6 replied:
Thank you Mike. This helps clarify things a lot!

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,181 @@
+\grace{I'm also confused why we have implementation of DA::NonNull but it doesn't do what the name says. Seems pretty counter intuitive.}
+
+
+\begin{lemma}[\texttt{DA::NonNull} contains \texttt{null}] \label{lemma-NonNull}
+\texttt{var} of type \texttt{DA::NonNull} can be of type \texttt{null}.
+\end{lemma}
+\begin{proof}
+Let the domain of atom variable \texttt{DA} be \texttt{InherentNullDomain<AllDomain<f64>>}. Recall that \texttt{InherentNullDomain} exists for types that can represent null inherently in the carrier type. Then the type $$\texttt{DA::NonNull == InherentNullDomain<AllDomain<f64>>::NonNull == f64}.$$ The latter holds because in the \texttt{InherentNullDomain} implementation in the rust code \url{htt ...
+
+Therefore \texttt{var} is also of type \texttt{f64}. \texttt{f64} can contain \texttt{null} values, so we are done.
+\end{proof}
```
3. is it correct that a variable with type DA::NonNull can contain null values?

Shoeboxam replied:
DA::NonNull may be capable of representing a null value. This is the case when DA is `InherentNullDomain<AllDomain<f64>>`, meaning DA::NonNull is f64. If an instance of DA::NonNull is null, it is either imputed or a construction-time error is raised by 2.

gracetian6 reviewed (commented):
Below I have marked down the questions and parts I'd like reviewers to prioritize.

Shoeboxam commented:
If it helps with clarity, we could rename `DA::NonNull` to `DA::Imputed` or something else. I'm indifferent towards renaming as it's not clear to me personally if that's an improvement.

Shoeboxam commented:
In the context of the overall library, this constructor slightly changes the meaning of `AllDomain<T>`. `AllDomain<T>` used to be "the set of values T may take on." It is now "the set of *non-null* values T may take on." This change to AllDomain simplifies the domains of all transformations/measurements after imputation, as it keeps descriptors for nullity isolated to two specific domains early in the data processing pipeline. On the other hand, this distinction is not currently captured in the member check for AllDomain. 

See also https://github.com/opendp/opendp/issues/203. If we adopted noisy-float, then we wouldn't be changing the AllDomain description.

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+
+\begin{itemize}
+    \item Either a valid \texttt{Transformation} is returned or an error is returned.
+    \item \grace{Should I also say that \texttt{impute\_constant} must satisfy the preconditions of make row by row? (See pseudo code \ref{line:row-transform}.) That is, \texttt{impute\_constant} must be a pure function?}
```
What is the best change for the post conditions?

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+
+\subsection{Proof}
+
+\begin{lemma}[\texttt{DA::NonNull} contains \texttt{null}] \label{lemma:NonNull}
+A \texttt{var} of type \texttt{DA::NonNull} can be of type \texttt{null}.
+\end{lemma}
```
Added this lemma per Mike's comments on why we must check constant for nullity in appropriate output domain proof outline.

SalilVadhan replied:
I don't follow what this lemma is saying.  it seems to be just a specific example (for f64), and don't want we want to say it *cannot* be of type null?

Shoeboxam replied:
Perhaps we should rename DA::NonNull to DA::Imputed. It represents the type after imputation. In the case of InherentNullDomain, DA::Imputed is a type like f64, which is a type that could be null. So we check that the constant is non-null at construction-time. Then in the function we check that each data element is non-null at runtime and replace with the known-to-be-non-null constant. 

The output of the function is always non-null, but it could be a type with the capacity to represent null.

gracetian6 replied:
Yes, I think renaming it to `Imputed` makes more sense

Shoeboxam replied:
It's been renamed in the repo.

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+Therefore \texttt{var} is also of type \texttt{f64}. \texttt{f64} can contain \texttt{null} values, so we are done.
+\end{proof}
+
+\begin{lemma}[Precondition for row transform]\label{lemma:pure-fn}
+\texttt{make\_impute\_constant} satisfies the preconditions of \texttt{make\_row\_by\_row}. That is, the atom function \texttt{impute\_constant} in pseudo code line \ref{line:atom-fn} is a pure function. 
+\end{lemma}
+\begin{proof}
```
Would like feedback on this proof for precondition for row transform (i.e. whether impute_constant helper function is a pure function)

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+\begin{proof}
+\begin{enumerate}
+
+\item \textbf{(Appropriate output domain).} 
+
+In the case of \texttt{make\_impute\_constant}, this corresponds to showing that for every vector $v$ of elements of type \texttt{DA::Carrier},  \texttt{function}$(v)$ is a vector of elements of type \texttt{DA::NonNull}. We can also say that \texttt{function}$(v)$ is a vector of elements that does not contain any \texttt{NonNull} values.
+
+The $\function(v)$ has type \texttt{Vec[DA::NonNull]} follows from the assumption that element $v$ is in \texttt{input\_domain} and from the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}). The type signature takes in an element of type \texttt{Vec[DA::Carrier]} and returns an element of type \texttt{Vec[DA::NonNull]}. If the Rust code co ...
+
+The type signature is not a sufficient check, since by Lemma \ref{lemma:NonNull}, the function's output type can represent a value (e.g. \texttt{float nan}) that is not a member in the \texttt{output\_domain}. \grace{How do I clarify that the output domain \\ \texttt{VectorDomain(AllDomain(DA::NonNull))} should actually contain NO NULLS? The code seems too loose because in Lemma 1.1 we just showe ...
```
Would like comments on appropriate output domain proof - the problem is that if we don't check constant for nullity, then the theorem won't work. The confusing part is output type `Vec[DA::NonNull]` may contain null values (counter-intuitive to the name `NonNull`) in Lemma 2.1, which is not part of the output domain `VectorDomain(AllDomain(DA::NonNull))` which should not contain null values. However, by the same logic of Lemma 2.1, that variable of type `DA::NonNull`, it seems like the output domain could technically contain null values.

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+
+We check whether the constant is null in pseudo code line \ref{line:null}, before it gets called in \texttt{function}. If \texttt{constant} is \texttt{null}, then the check raises a construction-time error, so the function is never run. Thus we are done.
+
+\grace{I didn't use the fact that v is of type DA::Carrier; am I doing something wrong?}
```
Minor concern

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+
+\item \textbf{(Stability guarantee).}     We know that $d_{in} \leq d_{out}$ because \texttt{Relation}($d_{in}$, $d_{out}$) $\texttt{ = True}$. Since the vectors $v, w$ are $d_{in}$-close, then $d_{Sym}(v, w) \leq d_{in}$.
+
+Recall that \texttt{make\_impute\_constant} satisfies the pure function precondition of \texttt{make\_row\_by\_row} in Lemma \ref{lemma:pure-fn} and \texttt{make\_impute\_constant} is a row transform (pseudocode line \ref{line:row-transform}). Then we can apply the stability guarantee of \href{https://www.overleaf.com/2489843918kjvyyytkjfsg}{\texttt{make\_row\_by\_row}}, which gives us $$d_{Sym}( ...
```
applied stability guarantee from row transform

gracetian6 reviewed (commented):
Changes as of Aug 5 (ready to be reviewed)
- stability guarantee pf references row transform
- proof of pure function (satisfies row transform criteria)
- row transform incorporated into pseudocode 

PRIORITIZE comments specifically added by me below!

SalilVadhan commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Variable \texttt{\texttt{constant}} must be of type \texttt{DA::NonNull}
+        \item Type \texttt{DA} must have traits \texttt{ImputableDomain}. % Unless the null method is defined, this means that \texttt{DA} is either \texttt{OptionNullDomain<AllDomain<T>>} or \texttt{InherentNullDomain<AllDomain<T>>}.
```
I don't see a definition of ImputableDomain in the pseudocode_defns file

SalilVadhan replied:
Actually, I found it - it is in the Domains section, not the traits section.  I think it's better to say that DA is a *domain* with the ImputableDomain trait, so that the reader knows where to look

SalilVadhan commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+        return d_out >= d_in*1
+
+    def function(data: Vec[DA::Carrier]) -> Vec[DA::NonNull]: |\label{line:fn}|
+        def impute_constant(x: DA) -> DA::NonNull: |\label{line:atom-fn}|
```
If impute_constant is defined as part of ImputableDomain, then it seems that this pseudocode should appear in the def of ImputableDomain.

SalilVadhan replied:
but you can repeat it in the proof below to remind the reader

SalilVadhan commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+            return constant if x.is_null else x
+        return list(map(impute_constant, data)) |\label{line:map}|
+        
+    # can we comment out input_metric, output_metric, and stability_relation, and function if it's not being called anymore?
```
yes it seems tahat we should remove the Relation from this pseudocode.  But in the last line impute_constant should probably be replaced with DA::impute_constant?

SalilVadhan commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+\texttt{make\_impute\_constant} satisfies the preconditions of \texttt{make\_row\_by\_row}. That is, the atom function \texttt{impute\_constant} in pseudo code line \ref{line:atom-fn} is a pure function. 
+\end{lemma}
+\begin{proof}
+We reference the examples in \href{https://en.wikipedia.org/wiki/Pure_function}{Pure function Wikipedia page} for the definition of a pure function, and later as a check list to verify the properties of a pure function hold. To verify whether \texttt{impute\_constant} is a pure function, it must satisfy the following properties:
```
we should not refer to the Wikipedia page, but rather include the definition of pure function we are using in one of our definitions documents

gracetian6 replied:
added in def 3.4 in proofs doc

SalilVadhan commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+    \item The function application has no side effects.
+\end{enumerate}
+
+The first property is satisfied because there is no static variable defined within \\\texttt{make\_impute\_constant}, no mutable reference argument, no return value of an input stream, all of which could potentially cause different function outputs for inputs. Even though the function returns a non-local variable \texttt{constant}, \texttt{constant} is never changed within the code of \texttt{mak ...
```
please ask Marco Gaboardi to review this when he comes back

SalilVadhan replied:
I don't think "constant" is a variable anymore once we've defined the function make_impute_constant; it is now hardwired as a constant in the function

SalilVadhan replied:
do you mean "none of which could produce side effects"?

SalilVadhan commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,196 @@
+The second property is satisfied because the local static variables and non-local variables are unchanged, and there are no mutable reference arguments or input/output streams called within the \texttt{impute\_constant}.
+\end{proof}
+
+\begin{theorem}
```
There should be no need for a new proof of this Theorem here.  Instead it should follow immediately from the corresponding theorem in the make_row_by_row, once you've established that the preconditions of make_row_by_row are met.

gracetian6 commented on impute_constant/impute_constant.tex:
```diff
@@ -0,0 +1,185 @@
+
+The $\function(v)$ has type \texttt{Vec[DA::Imputed]} follows from the assumption that element $v$ is in \texttt{input\_domain} and from the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}). The type signature takes in an element of type \texttt{Vec[DA::Carrier]} and returns an element of type \texttt{Vec[DA::Imputed]}. If the Rust code co ...
+
+The type signature is not a sufficient check, since by Lemma \ref{lemma:Imputed}, the function's output type can represent a value (e.g. \texttt{float nan}) that is not a member in the \texttt{output\_domain}. \grace{How do I clarify that the output domain \\ \texttt{VectorDomain(AllDomain(DA::Imputed))} should actually contain NO NULLS? The code seems too loose because in Lemma 1.1 we just showe ...
```
Would like feedback on how to clarify that **the output domain** `VectorDomain(AllDomain(DA::Imputed))` **should actually contain NO NULLS** but the output type `Vec[DA::Imputed]` can contain a null value? 

- Seems contradictory since  Lemma 1.1 we just showed that `DA::Imputed` can contain null values

gracetian6 commented:
New changes as of Sept 4, 2021:

- addressed comments about pure function (updated to proofs def doc)
- stylistic changes to pseudo code, pre conditions
- want feedback on appropriate output domain proof
